src/Orleans/Providers/ILeaseProvider.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.LeaseProviders
+{
+    /// <summary>
+    /// Acquired lease
+    /// </summary>
+    public class AcquiredLease 
+    {
+        /// <summary>
+        /// The resource key which the lease is attached to 
+        /// </summary>
+        public string ResourceKey { get; }
+        /// <summary>
+        /// Duration of the aquired lease
+        /// </summary>
+        public TimeSpan Duration { get; }
+        /// <summary>
+        /// Lease token, whcih will be null if acquiring or renewing the lease failed
+        /// </summary>
+        public string Token { get; }
+        /// <summary>
+        /// Start time for this lease, which is when the lease is acquired or renewed
+        /// </summary>
+        public DateTime StartTimeUtc { get; }
+    }
+
+    /// <summary>
+    /// AcquireLeaseResult class, which demonstrates result of acquiring or renewing lease operation
+    /// </summary>
+    public class AcquireLeaseResult
+    {
+        /// <summary>
+        /// Acquired lease, which will be null if acquire or renew operation failed.
+        /// </summary>
+        AcquiredLease AcquiredLease { get; }
+        /// <summary>
+        /// Whether successfully acquire or renew the lease
+        /// </summary>
+        public bool Success { get; }
+        /// <summary>
+        /// If acquiring or renewing the lease failed, this is the exception which caused it. This field would be null if operation succeed.
+        /// </summary>
+        public Exception FailureException { get; }
```

Comment:
I included ServiceFailure or unexpected errors into TransientFailure. In the view of the caller, service failure should be categorized to be retriable and transient. Eg, fix the service bug or misconfiguration is the service side's responsibility, once it is fixed in the service, the caller can retry on the service again.